File: day01/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <sstream>

#define IntVectorReference std::vector<int>

bool isDigit(const std::string &str)
{
    return !str.empty() && std::all_of(str.begin(), str.end(), ::isdigit);
}

std::pair<IntVectorReference, IntVectorReference> readFileLines()
{
    std::vector<int> left, right;
    //
    std::ifstream inputFile;
    inputFile.open("input.txt");
    std::string fileContent;
    std::string line;
    while (std::getline(inputFile, line))
    {
        std::istringstream iss(line);
        int num1, num2;

        if (iss >> num1 >> num2) 
        {
            left.push_back(num1);
            right.push_back(num2);
        }
        else
        {
            std::cerr << "Invalid line skipped: " << line << std::endl;
        }
    }
    std::sort(left.begin(), left.end());
    std::sort(right.begin(), right.end());
    return {left, right};
}

void part2(IntVectorReference left, IntVectorReference right)
{
    auto similarityScore = 0;
    for (const auto element: left) {
        similarityScore += std::count(right.begin(), right.end(), element) * element;
    }
    std::cout << "Part 2: " << similarityScore << std::endl;
}

void part1(IntVectorReference left, IntVectorReference right)
{
    auto distance = 0;
    for (std::size_t i = 0; i < left.size(); i++)
    {
        const int lvalue = left[i];
        const int rvalue = right[i];
        
        if (lvalue < rvalue)
            distance += rvalue - lvalue;
        else
            distance += lvalue - rvalue;
    }

    std::cout << "Part 1: " << distance << std::endl;
}


int main()
{
    const auto [left, right] = readFileLines();
    part1(left, right);
    part2(left, right);
    return 0;
}

File: day01/main.go
package main

import (
	"bufio"
	"os"
	"slices"
	"strconv"
	"strings"
)

func readLineByLine() []string {
	file, _ := os.Open("./input.txt")
	var lines []string
	defer file.Close()
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		lines = append(lines, line)
	}
	return lines
}

func diff(a, b int) int {
	if a < b {
		return b - a
	}
	return a - b
}

func part1(left, right []int) {
	distance := 0
	for i := 0; i < len(left); i++ {
		distance += diff(left[i], right[i])
	}
	println("Part 1:", distance)
}

func part2(left, right []int) {
	similarityScore := 0
	for _, element := range left {
		appearances := 0
		for _, rightElem := range right {
			if rightElem == element {
				appearances += 1
			}
		}
		similarityScore += element * appearances
	}
	println("Part 2:", similarityScore)
}

func main() {
	lines := readLineByLine()
	var left []int
	var right []int
	for _, element := range lines {
		slice := strings.Split(element, "   ")
		first, _ := strconv.Atoi(slice[0])
		last, _ := strconv.Atoi(slice[1])
		left = append(left, first)
		right = append(right, last)
	}
	slices.Sort(left)
	slices.Sort(right)
	//
	part1(left, right)
	part2(left, right)
}
